app/src/main/java/com/asma/tasky/feature_authentication/data/repository/AuthenticationRepositoryImpl.kt

```diff
@@ -0,0 +1,88 @@
+import com.asma.tasky.core.util.UiText
+import com.asma.tasky.feature_authentication.data.remote.AuthenticationApi
+import com.asma.tasky.feature_authentication.data.remote.request.LoginRequest
+import com.asma.tasky.feature_authentication.data.remote.request.RegistrationRequest
+import com.asma.tasky.feature_authentication.domain.repository.AuthenticationRepository
+import com.asma.tasky.feature_authentication.domain.util.AuthResult
+import retrofit2.HttpException
+import java.io.IOException
+
+class AuthenticationRepositoryImpl(
+    private val api: AuthenticationApi,
+    private val sharedPreferences: SharedPreferences
+) : AuthenticationRepository {
+
+    override suspend fun register(name: String, email: String, password: String): Resource<Unit> {
+        val registrationRequest =
+            RegistrationRequest(fullName = name, email = email, password = password)
+        return try {
+            val result = api.register(request = registrationRequest)
+            if (result.isSuccessful) Resource.Success(Unit)
+            else Resource.Error(UiText.DynamicString(result.message()))
+
+        } catch (e: IOException) {
+            Resource.Error(UiText.StringResource(R.string.netword_error))
+        } catch (e: HttpException) {
+            Resource.Error(UiText.StringResource(R.string.something_wrong))
+        }
+    }
+
+    override suspend fun login(email: String, password: String): Resource<Unit> {
+        val loginRequest = LoginRequest(email, password)
+        return try {
+            val result = api.login(loginRequest)
+            if (result.isSuccessful) {
+                result.body()?.let { response ->
+                    response.fullName
+                    println("JWT = ${response.token}")
+                    sharedPreferences.edit()
+                        .putString(Constants.KEY_JWT_TOKEN, response.token)
+                        .putString(Constants.KEY_USER_ID, response.userId)
+                        .apply()
+                }
+                Resource.Success(Unit)
+            } else
+                if (result.errorBody() != null)
+                    result.errorBody()!!.let {
+                        Resource.Error(UiText.DynamicString(it.string()))
+                    }
+                else Resource.Error(UiText.StringResource(R.string.unknown_error))
+
+        } catch (e: IOException) {
+            Resource.Error(UiText.StringResource(R.string.netword_error))
+        } catch (e: HttpException) {
+            Resource.Error(UiText.StringResource(R.string.something_wrong))
+        }
+    }
+
+    override suspend fun authenticate(): AuthResult {
+        return try {
+            val result = api.authenticate()
+            when (result.code()) {
+                200 -> AuthResult.Success
+                401 -> AuthResult.Unauthorized
+                else -> AuthResult.Error
+            }
+        } catch (e: Exception) {
+            AuthResult.Error
+        }
```

Comment:
Be careful with catching general exceptions inside suspend functions. You can break coroutine cancellation that way. If a coroutine is cancelled, the currently suspending function (here, it would be api.authenticate()) will throw a CancellationException. If your catch block eats that up, this exception won't be properly propagated up the coroutine scopes which can cause some coroutines to run longer than they should.
You can fix this by adding this:

```
if(e is CancellationException) throw e
```
or alternatively by only catching specific exceptions and not CancellationExceptions.

In this video I show the issue with that a bit more in detail (starting at 17:45):
https://www.youtube.com/watch?v=VWlwkqmTLHc